# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [110]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [111]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [112]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [113]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [114]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [115]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [116]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [117]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [118]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [119]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [120]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [121]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [122]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [123]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [124]:
data.duplicated().sum()

71

In [125]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [126]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [127]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [128]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [129]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [130]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Выведем из таблицы data необходимые данные для анализа, и сохраним в новую таблицу category, и переменуем названия столбцов для более наглядного представления:

In [131]:
category = data.rename(columns={'children': 'количеcтво детей',
                                'family_status': 'семейное положение',
                                'debt': 'заемщики',
                               'purpose_category': 'цели кридита',
                               'total_income_category': 'уровень дохода'})

print(category[['количеcтво детей','семейное положение',
                'заемщики','цели кридита','уровень дохода']])

       количеcтво детей семейное положение  заемщики  \
0                     1    женат / замужем         0   
1                     1    женат / замужем         0   
2                     0    женат / замужем         0   
3                     3    женат / замужем         0   
4                     0   гражданский брак         0   
...                 ...                ...       ...   
21520                 1   гражданский брак         0   
21521                 0    женат / замужем         0   
21522                 1   гражданский брак         1   
21523                 3    женат / замужем         1   
21524                 2    женат / замужем         0   

                   цели кридита уровень дохода  
0      операции с недвижимостью              B  
1        операции с автомобилем              C  
2      операции с недвижимостью              C  
3         получение образования              B  
4            проведение свадьбы              C  
...                         ...  

Теперь определим количество детей для каждого заемщика, а также список должников и найдем среднее значение:

In [132]:
table_children = category.groupby('количеcтво детей')['заемщики'].agg(['count','sum', 'mean'])
display(table_children)

count   sum      mean
количеcтво детей                       
0                 14091  1063  0.075438
1                  4808   444  0.092346
2                  2052   194  0.094542
3                   330    27  0.081818
4                    41     4  0.097561
5                     9     0  0.000000

Переведем среднее значение в проценты для более наглядного представления картины:

In [133]:
table_children['mean'] = table_children['mean'].apply(lambda x:f'{x:.2%}')
display(table_children)

count   sum   mean
количеcтво детей                    
0                 14091  1063  7.54%
1                  4808   444  9.23%
2                  2052   194  9.45%
3                   330    27  8.18%
4                    41     4  9.76%
5                     9     0  0.00%

Теперь переменуем столбцы:

In [134]:
table_children = table_children.rename(columns={'count': 'количество заемщиков',
                                                'sum': 'количество должников', 
                                                'mean': 'доля должников в %'})
display(table_children)

количество заемщиков  количество должников  \
количеcтво детей                                               
0                                14091                  1063   
1                                 4808                   444   
2                                 2052                   194   
3                                  330                    27   
4                                   41                     4   
5                                    9                     0   

                 доля должников в %  
количеcтво детей                     
0                             7.54%  
1                             9.23%  
2                             9.45%  
3                             8.18%  
4                             9.76%  
5                             0.00%

Отсортируем значения по таблице "доля должников в %" в порядке убывания:

In [135]:
table_children = table_children.sort_values('доля должников в %', ascending=False)
display(table_children)

количество заемщиков  количество должников  \
количеcтво детей                                               
4                                   41                     4   
2                                 2052                   194   
1                                 4808                   444   
3                                  330                    27   
0                                14091                  1063   
5                                    9                     0   

                 доля должников в %  
количеcтво детей                     
4                             9.76%  
2                             9.45%  
1                             9.23%  
3                             8.18%  
0                             7.54%  
5                             0.00%

**Вывод:** Больше всего задолжников по платежам прослеживается у представителей, имеющих одного, двух и четырех детей. У представителей, имеющих три и пять детей, наблюдаются более сниженные показатели, что может быть связанj с небольшим количеством клиентов, подпадающих под данную категорию.     

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Произведем группировку данных по семейному положению и возврату кредита в срок и также, как и в первом случае, рассчитаем список должников и их среднее значение:

In [136]:
family_table = category.groupby('семейное положение')['заемщики'].agg(['count','sum', 'mean'])
display(family_table)

count  sum      mean
семейное положение                         
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

Приведем таблицу в более наглядное представление:

In [137]:
# Приведем средние значения в процентное соотношение:
family_table['mean'] = family_table['mean'].apply(lambda x:f'{x:.2%}')

# Заменим значения в таблице:
family_table = family_table.rename(columns={'count': 'количество заемщиков',
                                            'sum': 'количество должников', 
                                            'mean': 'доля должников в %'})

# Отсортируем таблицу по доле должников в порядке убывания:
family_table = family_table.sort_values('доля должников в %', ascending=False)
display(family_table)

количество заемщиков  количество должников  \
семейное положение                                                  
Не женат / не замужем                  2796                   273   
гражданский брак                       4134                   385   
женат / замужем                       12261                   927   
в разводе                              1189                    84   
вдовец / вдова                          951                    63   

                      доля должников в %  
семейное положение                        
Не женат / не замужем              9.76%  
гражданский брак                   9.31%  
женат / замужем                    7.56%  
в разводе                          7.06%  
вдовец / вдова                     6.62%

**Вывод:** Не женатые и не замужние заемщики имеют наибольшую долю просрочек по кредиту. Высокие показатели также имеют заемщики состоящие в гражданском браке. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Теперь проверим зависимость от уровня дохода и возрата кридита в срок: 

In [138]:
# Определяем уровень дохода для каждого заемщика, а также список должников и находим среднее значение:
purpose_table = category.groupby('уровень дохода')['заемщики'].agg(['count','sum', 'mean'])

# Приведем средние значения в процентное соотношение:
purpose_table['mean'] = purpose_table['mean'].apply(lambda x:f'{x:.2%}')

# Заменим значения в таблице:
purpose_table = purpose_table.rename(columns={'count': 'количество заемщиков',
                                            'sum': 'количество должников', 
                                            'mean': 'доля должников в %'})

# Отсортируем таблицу по доле должников в порядке убывания:
purpose_table = purpose_table.sort_values('доля должников в %', ascending=False)
display(purpose_table)

количество заемщиков  количество должников доля должников в %
уровень дохода                                                               
E                                 22                     2              9.09%
C                              15921                  1353              8.50%
A                                 25                     2              8.00%
B                               5014                   354              7.06%
D                                349                    21              6.02%

**Вывод:** Наибольшая задолжность наблюдается у заемщиков, имеющих уровень дохода до 30 тысяч рублей, также большие значения имеют заемщики с доходом от 50 до 200 тысяч. Наименьшие показатели имеют заемщики с доходом от 200 тысяч до 1 млн рублей и от 30 до 50 тысяч рублей.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Теперь проверим, по каким целям кридита выявляется большая задолжность:

In [139]:
# Определяем цели кридита каждого заемщика, а также список должников и находим среднее значение:
income_table = category.groupby('цели кридита')['заемщики'].agg(['count','sum', 'mean'])

# Приведем средние значения в процентное соотношение:
income_table['mean'] = income_table['mean'].apply(lambda x:f'{x:.2%}')

# Заменим значения в таблице:
income_table = income_table.rename(columns={'count': 'количество заемщиков',
                                            'sum': 'количество должников', 
                                            'mean': 'доля должников в %'})

# Отсортируем таблицу по доле должников в порядке убывания:
income_table = income_table.sort_values('доля должников в %', ascending=False)
display(income_table)

количество заемщиков  количество должников  \
цели кридита                                                           
операции с автомобилем                    4279                   400   
получение образования                     3988                   369   
проведение свадьбы                        2313                   183   
операции с недвижимостью                 10751                   780   

                         доля должников в %  
цели кридита                                 
операции с автомобилем                9.35%  
получение образования                 9.25%  
проведение свадьбы                    7.91%  
операции с недвижимостью              7.26%

**Вывод:** Наибольшие задолжности наблюдатся по кридитам, связанным с образованием и операцимями с автомобилями. А самые низкие задолжности по кридиту наблюдаются по операциям с недвижимостью и провидением свадеб. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Самый основной - это человеческий фактор. Второй причиной могут быть отсутсвие данных, из-за чего могут возникать пропуски. Третьей причной может быть переформатирование данных, при котором теряются либо перезаписываются значения.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана показывает более объективную информацию, то есть не искажается  при больших и малых значений, тем самым обеспечивает лучшее представление о значениях

### Шаг 4: общий вывод.

Исходя из анализа всех таблиц можно сделать вывод, что больший приоритет при расмотрении заявки на кридит, будут иметь заемщики:
* Не имеющих детей и не состаящие в браке
* Имеющих доход от 30 до 50 тысяч рублей
* Желающие взять кридит на проведение свадьбы или операции, связанные с нидвижимостью

* Определяем количество детей для каждого заемщика, а также список должников и находим среднее значение в %:
* Отсортировывем таблицу по доле должников в порядке убывания:
* Получаем таблицу и проделываем тоже самое с уровнем дохода, с семейным положением и с цельями кридита:

| Количество детей                    | Количество заемщиков |количество должников | доля должников в % |
|--------------------------|-----------------|--------------------|-------------|
| 4               | 41                |     4         | 9.76       |
| 2               | 2052           | 194             |9.45       |
| 1               | 4808           | 444             | 9.23         |
| 3               | 330           | 27            | 8.18         |
| 0               | 14091           | 1063              | 7.54         |
| 5               | 9          | 0             | 0         |
